### PR TITLE
ROX-20936: Fix 'too_many_pings' in Compliance

### PR DESCRIPTION
**Pull request by vikin91**
## Description

This PR configures the keepalive params of the connection between Compliance (Node scanning) and Sensor, so that the client and server match and we no longer get  `too_many_pings` in the logs.

## Checklist
- [x] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

### Here I tell how I validated my change

1. I manually checked if the node scanning feature still works (in the UI).

2. Manual test on OCP:

- Regular deployment with `./deploy/openshift/deploy.sh`
- Setting compliance to spam sensor: `ocs set env daemonsets/collector --containers="compliance" ROX_NODE_SCANNING_INITIAL_BACKOFF="1s" ROX_NODE_SCANNING_MAX_INITIAL_WAIT="1s" ROX_NODE_SCANNING_INTERVAL="1m"`
- Observing Compliance logs for few hours (overnight)


Before (note the version):

```
│ No certificates found in /usr/local/share/ca-certificates                                                                                                                                                                                                            
│ No certificates found in /etc/pki/injected-ca-trust                                                                                                                                                                                                                
│ collection/compliance: 2024/01/08 16:42:54.655996 node_scanner.go:45: Info: Initialized gRPC connection to node-inventory container                                                                                                                                  
│ collection/compliance: 2024/01/08 16:42:54.656186 compliance.go:50: Info: Running StackRox Version: 4.3.x-631-gaf6b510092                                                                                                                                            
│ pkg/metrics: 2024/01/08 16:42:54.656260 server.go:142: Warn: Secure metrics server is disabled                                                                                                                                                                     
│ pkg/metrics: 2024/01/08 16:42:54.656342 server.go:125: Warn: Metrics server is disabled                                                                                                                                                                             
│ pkg/metrics: 2024/01/08 16:42:54.656381 server.go:142: Warn: Secure metrics server is disabled                                                                                                                                                                       │
│ pkg/metrics: 2024/01/08 16:42:54.656450 server.go:125: Warn: Metrics server is disabled                                                                                                                                                                              
│ pkg/metrics: 2024/01/08 16:42:54.656485 server.go:142: Warn: Secure metrics server is disabled                                                                                                                                                                       
│ collection/compliance: 2024/01/08 16:42:54.657782 compliance.go:65: Info: Initialized gRPC stream connection to Sensor                                                                                                                                               
│ collection/intervals: 2024/01/08 16:42:54.657925 intervals.go:64: Info: Scanning intervals: base interval: 1m0s, maximum absolute deviation from base: 1m0s, first scan starts not later than in: 1s                                                                 
│ collection/intervals: 2024/01/08 16:42:54.657952 intervals.go:72: Info: Initial scanning in 616.35134ms                                                                                                                                                              
│ collection/compliance: 2024/01/08 16:42:54.694135 compliance.go:282: Info: Successfully connected to Sensor at sensor.stackrox.svc:443                                                                                                                               
│ collection/compliance: 2024/01/08 16:42:55.274490 compliance.go:126: Info: Scanning node "piotr-01-08-99pc-rb7qj-worker-b-cfn6k"                                                                                                                                     
│ collection/intervals: 2024/01/08 16:42:57.612766 intervals.go:82: Info: Next node scan in 1m48.128418698s                                                                                                                                                            
│ collection/compliance: 2024/01/08 16:42:57.612841 compliance.go:126: Info: Scanning node "piotr-01-08-99pc-rb7qj-worker-b-cfn6k"                                                                                                                                     
│ collection/intervals: 2024/01/08 16:42:57.619450 intervals.go:82: Info: Next node scan in 41.598030849s                                                                                                                                                              
│ 2024/01/08 16:43:27 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".                                                                                                                      
│ collection/compliance: 2024/01/08 16:43:39.224050 compliance.go:126: Info: Scanning node "piotr-01-08-99pc-rb7qj-worker-b-cfn6k"                                                                                                                                     
│ collection/intervals: 2024/01/08 16:43:39.228174 intervals.go:82: Info: Next node scan in 1m28.993506076s                                                                                                                                                            
│ 2024/01/08 16:44:39 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".                                                                                                                      
│ collection/compliance: 2024/01/08 16:45:08.221857 compliance.go:126: Info: Scanning node "piotr-01-08-99pc-rb7qj-worker-b-cfn6k"                                                                                                                                     
│ collection/intervals: 2024/01/08 16:45:08.230182 intervals.go:82: Info: Next node scan in 1m54.239516394s
```

After:

```
No certificates found in /usr/local/share/ca-certificates
No certificates found in /etc/pki/injected-ca-trust
collection/compliance: 2024/01/08 16:48:07.047172 node_scanner.go:45: Info: Initialized gRPC connection to node-inventory container
collection/compliance: 2024/01/08 16:48:07.047320 compliance.go:50: Info: Running StackRox Version: 4.3.x-626-g78b7bd2397
pkg/metrics: 2024/01/08 16:48:07.047374 server.go:142: Warn: Secure metrics server is disabled
pkg/metrics: 2024/01/08 16:48:07.047441 server.go:125: Warn: Metrics server is disabled
pkg/metrics: 2024/01/08 16:48:07.047462 server.go:142: Warn: Secure metrics server is disabled
pkg/metrics: 2024/01/08 16:48:07.047519 server.go:125: Warn: Metrics server is disabled
pkg/metrics: 2024/01/08 16:48:07.047585 server.go:142: Warn: Secure metrics server is disabled
collection/compliance: 2024/01/08 16:48:07.048593 compliance.go:65: Info: Initialized gRPC stream connection to Sensor
collection/intervals: 2024/01/08 16:48:07.048657 intervals.go:64: Info: Scanning intervals: base interval: 1m0s, maximum absolute deviation from base: 1m0s, first scan starts not later than in: 1s
collection/intervals: 2024/01/08 16:48:07.048673 intervals.go:72: Info: Initial scanning in 532.978803ms
collection/compliance: 2024/01/08 16:48:07.086540 compliance.go:282: Info: Successfully connected to Sensor at sensor.stackrox.svc:443
collection/compliance: 2024/01/08 16:48:07.581801 compliance.go:126: Info: Scanning node "piotr-01-08-99pc-rb7qj-worker-c-mn8cf"
collection/intervals: 2024/01/08 16:48:09.779830 intervals.go:82: Info: Next node scan in 55.857394293s
collection/compliance: 2024/01/08 16:48:09.779881 compliance.go:126: Info: Scanning node "piotr-01-08-99pc-rb7qj-worker-c-mn8cf"
collection/intervals: 2024/01/08 16:48:09.783992 intervals.go:82: Info: Next node scan in 1m15.620288609s
collection/compliance: 2024/01/08 16:49:25.407523 compliance.go:126: Info: Scanning node "piotr-01-08-99pc-rb7qj-worker-c-mn8cf"
collection/intervals: 2024/01/08 16:49:25.412079 intervals.go:82: Info: Next node scan in 13.335713363s

```

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: pkg/clientconn/client.go [hunk 1]
```diff
@@ -285,7 +285,14 @@ func AuthenticatedGRPCConnection(endpoint string, server mtls.Subject, extraConn
...
 
 	var dialOpts []grpc.DialOption
-	dialOpts = append(dialOpts, keepAliveDialOption())
+	// To avoid getting 'Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings"'
+	// from Sensor in Compliance, we must match the client and server settings for keepalive
+	// See: https://github.com/grpc/grpc/blob/master/doc/keepalive.md#faq
+	dialOpts = append(dialOpts, grpc.WithKeepaliveParams(keepalive.ClientParameters{
+		Time:                10 * time.Second,
+		Timeout:             40 * time.Second,
+		PermitWithoutStream: false,
```
Comment (lvalerom) [inline pkg/clientconn/client.go:294]: Just out of curiosity; why did you change `PermitWithoutStream` to `false`? Couldn't this lead to problems in some scenarios?

Comment (vikin91) [reply pkg/clientconn/client.go:294]: That was the primary reason for the error. In https://github.com/grpc/grpc/blob/master/doc/keepalive.md#faq, we can read:

> if a server has GRPC_ARG_KEEPALIVE_PERMIT_WITHOUT_CALLS set to false while the client has set this to true resulting in keepalive pings being sent even when there is no call in flight.

Compliance was a client here, so I changed the setting to see if that helps - it does. I think, we have a streams between: (1) Sensor and Central, (2) Collector and Sensor, but not between Compliance and Sensor.

Second factor was:

> if the client's GRPC_ARG_KEEPALIVE_TIME_MS setting is lower than the server's GRPC_ARG_HTTP2_MIN_RECV_PING_INTERVAL_WITHOUT_DATA_MS.

Which I think is the `10s` parameter in that block of code. It was okay, but to be double sure, I changed the timeout from 30s to 40s to match the settings in sensor.

**Review by lvalerom — APPROVED** [review]

File: pkg/clientconn/client.go (hunk 1 above)
Comment (lvalerom) [reply pkg/clientconn/client.go:294]: Gotcha! That makes sense thank you 🙂
